extension/agenthealth/handler/stats/provider/statuscode.go

```diff
@@ -0,0 +1,216 @@
+func (sp *StatsProvider) processStatusCode(entry statusCodeEntry) {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+
+	stats, exists := sp.statsByOperation[entry.operation]
+	if !exists {
+		stats = &[5]int{}
+		sp.statsByOperation[entry.operation] = stats
+	}
+	sp.updateStatusCodeCount(stats, entry.statusCode)
+}
+
+// updateStatusCodeCount updates the count for the specific status code.
+func (sp *StatsProvider) updateStatusCodeCount(stats *[5]int, statusCode int) {
+	log.Printf("Updating stats for status code: %d\n", statusCode)
+	switch statusCode {
+	case 200:
+		stats[0]++
+	case 400:
+		stats[1]++
+	case 408:
+		stats[2]++
+	case 413:
+		stats[3]++
+	case 429:
+		stats[4]++
+	default:
+		log.Printf("Ignored status code: %d\n", statusCode)
+	}
+}
+
+// startResetTimer initializes a reset timer to clear stats periodically.
+func (sp *StatsProvider) startResetTimer() {
+	log.Println("Starting stats reset timer...")
+	sp.resetTimer = time.AfterFunc(statusResetInterval, func() {
+		sp.mu.Lock()
+		defer sp.mu.Unlock()
+		log.Println("Resetting stats...")
+		for key := range sp.statsByOperation {
+			delete(sp.statsByOperation, key)
+		}
+		sp.startResetTimer()
+	})
+}
+
+// StatusCodeHandler is the handler that uses the StatsProvider for processing.
+type StatusCodeHandler struct {
+	statsProvider *StatsProvider
+}
+
+func (h *StatusCodeHandler) ID() string {
+	return statusHandlerID
+}
+
+func (h *StatusCodeHandler) Position() awsmiddleware.HandlerPosition {
+	return awsmiddleware.After
+}
+
+// NewStatusCodeHandler creates a new handler with the given StatsProvider.
+func NewStatusCodeHandler(provider *StatsProvider) *StatusCodeHandler {
+	log.Println("Creating new StatusCodeHandler...")
+	return &StatusCodeHandler{statsProvider: provider}
+}
+
+// HandleResponse enqueues the status code into the StatsProvider's channel.
+func (h *StatusCodeHandler) HandleResponse(ctx context.Context, r *http.Response) {
+	operation := awsmiddleware.GetOperationName(ctx)
+	if !h.statsProvider.filter.IsAllowed(operation) {
+		return
+	}
+
+	operation = GetShortOperationName(operation)
+	if operation == "" {
+		return
+	}
+
+	h.statsProvider.EnqueueStatusCode(operation, r.StatusCode)
+}
+
+// GetShortOperationName maps long operation names to short ones.
+func GetShortOperationName(operation string) string {
+	switch operation {
+	case "PutRetentionPolicy":
+		return "pmd"
+	case "DescribeInstances":
+		return "di"
+	case "DescribeTags":
+		return "dt"
+	case "DescribeTasks":
+		return "dts"
+	case "DescribeVolumes":
+		return "dv"
+	case "DescribeContainerInstances":
+		return "dci"
+	case "DescribeServices":
+		return "ds"
+	case "DescribeTaskDefinition":
+		return "dtd"
+	case "ListServices":
+		return "ls"
+	case "ListTasks":
+		return "lt"
+	case "CreateLogGroup":
+		return "clg"
+	case "CreateLogStream":
+		return "cls"
+	default:
+		return ""
+	}
+}
+
+// Stats returns the aggregated stats for operations.
+func (sp *StatsProvider) Stats(_ string) agent.Stats {
```

Comment:
Right now, this means that each call to `Stats` is going to include the current status code map. This function is called for each request that goes through the `StatsHandler` since it needs to add the data to the header, which means that we're going to get duplicate counts during that time period.

https://github.com/aws/amazon-cloudwatch-agent/blob/a8ef6f778b075e3eaf51e033a15313edaca33d89/extension/agenthealth/handler/stats/handler.go#L72-L74

For example, if we record that "DescribeTags" had a 200, each PMD/PLE/PTS call made until the reset timer goes off would have that count.

Reply:
Ahh that's right, we need to reset the count that has already been sent so:

statuscode records DescribeTags with a 200: {"dt":[1,0,0,0,0]}
Reset sent map.
PMD sends the statuscodemap as part of the stats header
statuscode records DescribeTags with a 200: statuscodemap{"dt":[1,0,0,0,0]}
PMD sends the statuscodemap again
We'll record 2 200s for DT